tools/doc/apilinks.js

```diff
@@ -0,0 +1,163 @@
+  hash = child_process.execSync(
+    `git describe --contains ${hash}`,
+    { stdio: ['ignore', null, 'ignore'] }
+  ).toString();
+} catch {
+}
+
+const definition = {};
+process.argv.slice(2).forEach((file) => {
+  const basename = path.basename(file, '.js');
+
+  // parse source
+  const source = fs.readFileSync(file, 'utf8');
+  const ast = acorn.parse(source, { ecmaVersion: 10 });
+  const program = ast.body;
+
+  // scan for line termination characters
+  const eol = /\n/g;
+  let match;
+  const lineends = [0];
+  while (match = eol.exec(source)) {
+    lineends.push(match.index);
+  }
+
+  // Binary search lineends to find the line number.  Note that line numbers
+  // start with one (that's why lineends has an extra zero prepended to the
+  // list), and that the end of the loop, min and max point to the line ends
+  // immediately before and immediately after the indicated position.
+  function lineno(position) {
+    var min = 0;
+    var max = lineends.length;
+    while (max - 1 > min) {
+      const mid = Math.floor((min + max) / 2);
+      if (position > lineends[mid]) {
+        min = mid;
+      } else {
+        max = mid;
+      }
+    }
+    return max;
+  }
+
+  // scan for exports
+  const exported = { constructors: [], identifiers: [] };
+  program.forEach((statement) => {
+    if (statement.type !== 'ExpressionStatement') return;
+    const expr = statement.expression;
+    if (expr.type !== 'AssignmentExpression') return;
+
+    let lhs = expr.left;
+    if (expr.left.object.type === 'MemberExpression') lhs = lhs.object;
+    if (lhs.type !== 'MemberExpression') return;
+    if (lhs.object.name !== 'module') return;
+    if (lhs.property.name !== 'exports') return;
+
+    let rhs = expr.right;
+    while (rhs.type === 'AssignmentExpression') rhs = rhs.right;
+
+    if (rhs.type === 'NewExpression') {
+      exported.constructors.push(rhs.callee.name);
+    } else if (rhs.type === 'ObjectExpression') {
+      rhs.properties.forEach((property) => {
+        if (property.value.type === 'Identifier') {
+          exported.identifiers.push(property.value.name);
+          if (/^[A-Z]/.test(property.value.name[0])) {
+            exported.constructors.push(property.value.name);
+          }
+        }
+      });
+    } else if (rhs.type === 'Identifier') {
+      exported.identifiers.push(rhs.name);
+    }
+  });
+
+  // scan for definitions matching those exports; currently supports:
+  //
+  //   ClassName.foo = ...;
+  //   ClassName.prototype.foo = ...;
+  //   function Identifier(...) {...};
+  //
+  const link = `https://github.com/${repo}/blob/${hash}/${file}`;
+  program.forEach((statement) => {
+    if (statement.type === 'ExpressionStatement') {
+      const expr = statement.expression;
+      if (expr.type !== 'AssignmentExpression') return;
+      if (expr.left.type !== 'MemberExpression') return;
+
+      let object;
+      if (expr.left.object.type === 'MemberExpression') {
```

Comment:
Another ignorable performance nit: maybe we should cache such things more; for example, we retrieve `expr.left.object` many times here. But I am not sure if the gain is worth to create medium variables.